rfcs/0000-cfb-encoding/0000-cfb-encoding.md

```diff
@@ -0,0 +1,98 @@
+# CFB Encoding
+
+CFB (Canonical FlatBuffers) is a restricted variant of FlatBuffers for producing unequivocal transfer syntax.
+
+As described in [FlatBuffer Internals](#)
+
+> On purpose, the format of FlatBuffers leaves a lot of details about where exactly things live in memory undefined, e.g. fields in a table can have any order, and objects to some extent can be stored in many orders. This is because the format doesn't need this information to be efficient, and it leaves room for optimization and extension (for example, fields can be packed in a way that is most compact). Instead, the format is defined in terms of offsets and adjacency only. This may mean two different implementations may produce different binaries given the same input values, and this is perfectly valid.
+
+CFB eliminates all the undefined encodings. For a given data structure, the encoded binary is determined.
+
+CFB is useful in cryptography to generate a determined hash from a complex data structure:
+
+1.  Encode the data structure into binary using CFB.
+2.  Hash the binary as the identifier.
+
+The identifier generated by original FlatBuffers is malleable. Since it is easy to encode the data into a different binary without modifying the data itself.
+
+The name comes from [CER](#), which is a restricted variant of BER to encode data structures described by ASN.1.
+
+## The Algorithm
+
+The buffer consists of 3 portions:
+
+- root table pointer
+- vtable portion
+- data portion
+
+Root table pointer is a 4 bytes little endian unsigned integer, which equals the start position of the data portion.
```

Comment:
Does this mean that the serialized data cannot exceed 4GB? While this won't be a problem now, note that in CKB one cell could grow to 4GB by itself, that will make the enclosing transaction or block bigger than 4GB. Maybe it's time to take this opportunity and bump this to 8GB for future proof reason?

Reply:
No, it just means the vtable portion is limited to less than 4G.

The problem is that the Vector and String length is encoded with 4 bytes. A simple solution is using a list of `[ubyte]` to store data.